util/io_posix.cc

```diff
@@ -58,124 +58,45 @@ const size_t kPageSize = sysconf(_SC_PAGESIZE);
   return uintptr_t(ptr) % (kPageSize) == 0;
 }
 
-Status ReadAligned(int fd, Slice* data, const uint64_t offset,
-                   const size_t size, char* scratch) {
-  assert(IsSectorAligned(offset));
-  assert(IsSectorAligned(size));
-  assert(IsPageAligned(scratch));
-
-  size_t bytes_read = 0;
-  ssize_t status = -1;
-  while (bytes_read < size) {
-    status =
-        pread(fd, scratch + bytes_read, size - bytes_read, offset + bytes_read);
-    if (status <= 0) {
-      if (errno == EINTR) {
-        continue;
-      }
-      break;
-    }
-    bytes_read += status;
-    if (status % static_cast<ssize_t>(kSectorSize) != 0) {
-      // Bytes reads don't fill sectors. Should only happen at the end
-      // of the file.
-      break;
-    }
-  }
-
-  *data = Slice(scratch, bytes_read);
-  return status < 0 ? Status::IOError(strerror(errno)) : Status::OK();
 }
 
-Status ReadUnaligned(int fd, Slice* data, const uint64_t offset,
-                     const size_t size, char* scratch) {
-  assert(scratch);
-  assert(!IsSectorAligned(offset) || !IsSectorAligned(size) ||
-         !IsPageAligned(scratch));
-
-  const uint64_t aligned_off = Lower(offset, kSectorSize);
-  const size_t aligned_size = Upper(size + (offset - aligned_off), kSectorSize);
-  auto aligned_scratch = NewAligned(aligned_size);
-  assert(aligned_scratch);
-  if (!aligned_scratch) {
-    return Status::IOError("Unable to allocate");
-  }
-
-  assert(IsSectorAligned(aligned_off));
-  assert(IsSectorAligned(aligned_size));
-  assert(aligned_scratch);
-  assert(IsPageAligned(aligned_scratch.get()));
-  assert(offset + size <= aligned_off + aligned_size);
-
-  Slice scratch_slice;
-  Status s = ReadAligned(fd, &scratch_slice, aligned_off, aligned_size,
-                         reinterpret_cast<char*>(aligned_scratch.get()));
-
-  // copy data upto min(size, what was read)
-  memcpy(scratch, reinterpret_cast<char*>(aligned_scratch.get()) +
-                      (offset % kSectorSize),
-         std::min(size, scratch_slice.size()));
-  *data = Slice(scratch, std::min(size, scratch_slice.size()));
-  return s;
-}
-
-Status DirectIORead(int fd, Slice* result, size_t off, size_t n,
-                    char* scratch) {
-  if (IsSectorAligned(off) && IsSectorAligned(n) && IsPageAligned(scratch)) {
-    return ReadAligned(fd, result, off, n, scratch);
-  }
-  return ReadUnaligned(fd, result, off, n, scratch);
-}
-}  // namespace
-
 /*
  * PosixSequentialFile
  */
-PosixSequentialFile::PosixSequentialFile(const std::string& fname, FILE* f,
-                                         const EnvOptions& options)
+PosixSequentialFile::PosixSequentialFile(const std::string& fname, FILE* file,
+                                         int fd, const EnvOptions& options)
     : filename_(fname),
-      file_(f),
-      fd_(fileno(f)),
-      use_direct_io_(options.use_direct_reads) {}
+      file_(file),
+      fd_(fd),
+      direct_io_(options.use_direct_reads && !options.use_mmap_reads) {}
```

Comment:
Is it a configuration mistake to have use_direct_reads == true and use_mmap_reads == true? If so, can we add an assertion somewhere instead of silently disabling direct I/O?